### PR TITLE
Allow multiple certs for one (main) domain

### LINKED ISSUE #40: [ Feature Request ] Allow the same certificate on multiple nodes
I am in the process of migrating a service from one node to another. As part of this, I set up the same certificates on the second node, both with the respective hostname as a SAN:

The puppet run completes on the second node, but on the puppet master it now errors out with:
```
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Resource Statement, Duplicate declaration: Acme::Request[myservice.myhost.tld] is already declared; cannot redeclare (file: /etc/puppetlabs/code/environments/production/modules/acme/manifests/csr.pp, line: 192) (file: /etc/puppetlabs/code/environments/production/modules/acme/manifests/csr.pp, line: 192, column: 5) (file: /etc/puppetlabs/code/environments/production/modules/acme/manifests/certificate.pp, line: 59) on node puppetmaster.myhost.tld
```

The certificates are set up in hiera like this:

on node1 (which is also the puppetmaster, in case that matters):
```
acme::certificates:
 'myservice.myhost.tld node1.myhost.tld':
    use_profile: my_profile
```

on node2:
```
acme::certificates:
 'myservice.myhost.tld node2.myhost.tld':
    use_profile: my_profile
```

Perhaps the acme::request resources that are exported "to" the puppetmaster should be canonicalized with the hostname?

### PR DESCRIPTION
**Pull request by oxc**
Before, only one certificate could exist for a specific (main) domain
across all nodes. This change allows using the $name of the resources to
give the certificate a different identifier, and specify the domain(s)
as $domain.

Fixes #40

This change replaces the usage of $domain as resource title for nearly everything. It uses the resource name also for child resources.
If the name is different from the domain, it will specify a different `--cert-home` to acme.sh, which is only used by this particular certificate. For this to work, the CERT_NAME property has to be removed from the account config.
The single facts for each certificate are replace by two structured facts that contain all resources. This makes it easier to use the unaltered resource names without having to escape characters. The acme_csrs fact is a hash from certificate name to csr content, the acme_certs fact is a hash from certificate name to a hash containing { crt, ca, cn }.

For all certificates that don't explicitly set the `domain` property, as well as for single domain certificates, this should be fully compatible. However, for SAN certificates that specify the `domain` as well, this is currently a breaking change, as they will use new directory names (based on their $name, which is the full domain list). They should get fully regenerated, but the old files including the private key will be lost if not migrated at some point.

### CODE REVIEW

**Comment by oxc** [comment]
@fraenki, I've tried to implement the idea I outlined in #40. This is completely untested, except for the facts, but I wanted to get your opinion early, because I changed a lot of internal workings.

I will test the changes tomorrow, looking forward to your feedback.

File: manifests/certificate.pp [hunk 1]
```diff
@@ -40,23 +43,35 @@
...
+    # compatibility mode, parse name as list of domains, and use first as certificate resource name
+    $domains = split(downcase($name), ' ')
+    $cert_name = $domains[0]
+  } elsif $domain =~ String {
+    $domains = split($domain, ' ')
+    $cert_name = $name
+  } else {
+    $domains = $domain
+    $cert_name = $name
+  }
```
Comment (oxc) [inline manifests/certificate.pp:58]: I found a way to make this backwards-compatible after all. As long as the domain property of a certificate is not set, the name of the created resources will be the main certificate, as it was before.

Only if a certificate was created with both the resource name and the domain property set to a list of domains will this be a breaking change.

**Comment by oxc** [comment]
I've tested this no my setup and made some more minor adjustments. All my existing certificates are kept as-is without any modifications, and new ones are created correctly, even when specifying a custom resource name (and even if that resource name contains spaces and parenthesis 😊)

The crt resources are now collected based on their title, this allows us to fully get rid of the wildcard marker.

**Comment by oxc** [comment]
I don't like how you still have to make sure that your certificate names are globally unique across nodes. However, I don't see a good way to introduce automatic hostname scoping without breaking compatibility. Any ideas?

**Comment by fraenki** [comment]
> However, I don't see a good way to introduce automatic hostname scoping without breaking compatibility. Any ideas?

Unfortunately no. Maybe this is another feature that could be hidden behind an opt-in flag? Enabling the flag would break backwards compatibility, but allow the same cert on multiple nodes.

**Review by fraenki — APPROVED** [review]

**Review by fraenki — APPROVED** [review]

**Comment by fraenki** [comment]
Merged, thanks!

**Pull request merged**